### PR TITLE
Consider the StringBuilder's capacity instead of content length when being trimmed

### PR DESCRIPTION
**Pull request by xnslong**
The trim operation aims at releasing the too much memory occupied by the `StringBuilder`. So when the StringBuilder really contains few effective characters, but with large capacity, should also be trimmed.

### CODE REVIEW

**Comment by xnslong** [comment]
@garydgregory Yes, I noticed that and fixed it.

File: log4j-core/src/test/java/org/apache/logging/log4j/core/layout/AbstractStringLayoutTest.java [hunk 1]
```diff
@@ -59,12 +59,14 @@ public void testGetStringBuilderCapacityRestrictedToMax() throws Exception {
         final int LARGE = 4096;
         final String largeMessage = new String(new char[LARGE]);
         final StringBuilder sb2 = ConcreteStringLayout.getStringBuilder();
-        assertEquals("resized capacity", GROWN, sb2.capacity());
+        assertTrue("reset capacity", sb2.capacity() <= AbstractStringLayout.MAX_STRING_BUILDER_SIZE);
```
Comment (garydgregory) [inline log4j-core/src/test/java/org/apache/logging/log4j/core/layout/AbstractStringLayoutTest.java]: Can't we have an exact comparison here? This feels hacky to me :-(

**Comment by garydgregory** [comment]
Hi,

When I build with 'mvn clean install' from a clean git master, all the tests pass. Note that I am on Windows 10 and that the test HttpAppenderTest fails randomly for me, so I usually annotate it with @Ignored in my local workspace. It looks like you have a debugging session ahead of you in your favorite IDE ;-)

**Comment by xnslong** [comment]
Well, the case runs well on my computer. But I notice that the `MAX_STRING_BUILDER_SIZE` is not a constant, instead it may be loaded from the sysstem properties. So I changed my test case, to define the size upon the value of `MAX_STRING_BUILDER_SIZE`, instead of a hard coded number. Let's see, whether it will work.

**Comment by xnslong** [comment]
Still reports this problem? Hmm... interesting. I also have ran this case in my IDE, nothing went wrong. Would you please show me the git repository version, maybe we have to check if we are running the same version of code.

**Comment by xnslong** [comment]
I merged the latest master to my branch just now, Would you please run the case again, and show me the result? @garydgregory

**Comment by xnslong** [comment]
I reappeared the failure! when I run the tests with `mvn clean test -pl log4j2-core`, I got the following failure (it is just the failure you provided)

```
[ERROR] Failures: 
[ERROR]   AbstractStringLayoutTest.testGetStringBuilderCapacityRestrictedToMax:74 capacity, trimmed to MAX_STRING_BUILDER_SIZE expected:<2048> but was:<4096>
```

Then I run mvn in debug mode, and find out the `log4j2-api` is not my local code, but some code downloaded from the maven snapshot repository. 

> command for running test in debug mode:
> 
> ```
> mvn test -pl log4j-core -Dmaven.surefire.debug="-Xdebug -Xrunjdwp:transport=dt_socket,server=y,suspend=y,address=8000 -Xnoagent -Djava.compiler=NONE" -Dtest=AbstractStringLayoutTest
> ```

Well, the correct command for running the test is `mvn clean test -am -pl log4j2-core`. It will compile `log4j2-core` and all its dependent modules, but I didn't have java9 installed, so it will always fail for the `log4j-api-java9`. If you have java9 installed, please check it to see whether it fails.

**Comment by garydgregory** [comment]
I have the whole project sources and all modules (sources) loaded in
Eclipse so I run the one test directly from Eclipse.

The commands you show will never work (did you try them?) since there is no
log4j2-core module, it is log4j-core.

So if I run:

mvn clean install -pl log4j-api

and:

mvn clean test -pl log4j-core -Dtest=AbstractStringLayoutTest

I get:

[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.logging.log4j.core.layout.AbstractStringLayoutTest
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0, Time elapsed:
0.032 s <<< FAILURE! - in
org.apache.logging.log4j.core.layout.AbstractStringLayoutTest
[ERROR]
testGetStringBuilderCapacityRestrictedToMax(org.apache.logging.log4j.core.layout.AbstractStringLayoutTest)
 Time elapsed: 0.032 s  <<< FAILURE!
java.lang.AssertionError: capacity, trimmed to MAX_STRING_BUILDER_SIZE
expected:<2048> but was:<4096>
        at
org.apache.logging.log4j.core.layout.AbstractStringLayoutTest.testGetStringBuilderCapacityRestrictedToMax(AbstractStringLayoutTest.java:74)

[INFO]
[INFO] Results:
[INFO]
[ERROR] Failures:
[ERROR]
AbstractStringLayoutTest.testGetStringBuilderCapacityRestrictedToMax:74
capacity, trimmed to MAX_STRING_BUILDER_SIZE expected:<2048> but was:<4096>
[INFO]
[ERROR] Tests run: 1, Failures: 1, Errors: 0, Skipped: 0
[INFO]
[INFO]
------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO]
------------------------------------------------------------------------
[INFO] Total time: 53.099 s
[INFO] Finished at: 2017-07-17T08:53:05-07:00
[INFO] Final Memory: 43M/686M

Make sure you have the latest from git master.

Thank you!
Gary

**Comment by xnslong** [comment]
After this failure, could you please check your local maven repository for the log4j-api-2.9-SNAPSHOT version, decompile the jar and check the StringBuilders class. After I got the failure, I found it is not compiled from my code.

I didn't check the command I provided above, I just wrote it on my experience. My point is that the two modules may have to be compiled in only 1 command. For me, my settings for maven will download the dependent SNAPSHOT artifacts every time I build a module, so even when I installed the SNAPSHOT version, it will be rewritten when I try to build the other module. I'm not sure what's your settings, so I suggest to check the jar file in the repository.

```java
public static void trimToMaxSize(final StringBuilder stringBuilder, final int maxSize) {
    if (stringBuilder != null && stringBuilder.capacity() > maxSize) { // I modified this line, but I found it remains unchanged and still check the stringBuilder.length() in the decompiled code.
        stringBuilder.setLength(maxSize);
        stringBuilder.trimToSize();
    }
}
```

**Comment by xnslong** [comment]
I followed the tips to run command `mvn -U clean test -pl log4j-core -Dtest=AbstractStringLayoutTest`, and then I checked my local maven repository, but I'm afraid it didn't work. The following picture is a screen shot when I decompile the `log4j-api-2.9-SNAPSHOT.jar` with `Java Decompiler`. Please have a look on the red rectangle I draw here, it is where different from my code, and also the reason cause the failures above. 

screen shot of decompiling the log4j-api-2.9-SNAPSHOT.jar

**Comment by garydgregory** [comment]
Please accept my apologies, I made a mistake saying the patch was partly applied to log4j-api. It was not. The new test passes.

**Comment by garydgregory** [comment]
With this patch, I now get:

[ERROR] Failures:
[ERROR]   AsyncLoggerThreadContextGarbageFreeTest>AbstractAsyncThreadContextTestBase.testAsyncLogWritesToLog:159->AbstractAsyncThreadContextTestBase.checkResult:185 AsyncLoggerTest.log: line 18 expected:<...yncLoggerContext i=1[]8> but was:<...yncLoggerContext i=1[2]8>
[INFO]
[ERROR] Tests run: 1889, Failures: 1, Errors: 0, Skipped: 21
[INFO]
[INFO] ------------------------------------------------------------------------

Gary

**Comment by xnslong** [comment]
I runned this test, but it didn't fail on my machine.

**Pull request closed without merging**